File: matlab/process_and_simulation_data.m
switch exampleID
    case 0 % Deafult example - Simple code with noise and filtering

        %%%%%% Process parameters        
        K=1;            % Static gain
        T=1;            % Time constant
        L=0.5;          % Time delay
        
        y0=0;           % Initial value of the process ouput
        u0=0;           % Initial value of the controller ouput
        
        %%%%%% Simulation paramters
        Tsim=20+L;      % Simulation time  
        
        Dt=0.01;        % Sampling time
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        uc=zeros(Tsim/Dt,1);    % Control signal array for b=1
        yc=zeros(Tsim/Dt,1);    % Process output array for b=1
        rc=zeros(Tsim/Dt,1);     % Reference signal array
        r=0;                     % Initial reference value
        
        % Noise signal
        SNRdb = 5; %Given SNRdb
        variance = 1/(Dt*(10^(SNRdb/10))); 
        load semilla
        rng(s); 
        W = sqrt(variance).*randn(Tsim/Dt); %Gaussian white noise W
        noisesignal=0.01*W;

    case 1 % Example 1 - From man to auto

        %%%%%% Process parameters
        
        K=1;            % Static gain
        T=1;            % Time constant
        L=0.5;          % Time delay
        
        y0=0;           % Initial value of the process ouput
        u0=0;           % Initial value of the controller ouput
        
        %%%%%% Simulation paramters
        Tsim=20+L;      % Simulation time  
        
        Dt=0.01;        % Sampling time
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        uc=zeros(Tsim/Dt,1);    % Control signal array for b=1
        yc=zeros(Tsim/Dt,1);    % Process output array for b=1
        rc=zeros(Tsim/Dt,1);     % Reference signal array
        r=3;                     % Initial reference value    

    case 2 % Example 2 - P, PD control and rate limitation

        %%%%%% Process parameters
        
        K=1;            % Static gain
        T=1;            % Time constant
        L=0.5;          % Time delay
        
        y0=0;           % Initial value of the process ouput
        u0=1;           % Initial value of the controller ouput
        
        %%%%%% Simulation paramters
        Tsim=30+L;      % Simulation time  
        
        Dt=0.1;        % Sampling time
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        uc=ones(Tsim/Dt,1);     % Control signal array for b=1
        yc=ones(Tsim/Dt,1);     % Process output array for b=1
        rc=ones(Tsim/Dt,1);     % Reference signal array
        r=1;                    % Initial reference value            

    case 3 % Example 3 - Set-point weighting

        %%%%%% Process parameters
        
        K=1;            % Static gain
        T=1;            % Time constant
        L=0.5;          % Time delay
        
        y0=0;           % Initial value of the process ouput
        u0=0;           % Initial value of the controller ouput
        
        %%%%%% Simulation paramters
        Tsim=20+L;   % Simulation time  

        Dt=0.01;        % Sampling time        
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        uc1=zeros(Tsim/Dt,1);    % Control signal array for b=1
        uc2=zeros(Tsim/Dt,1);    % Control signal array for b=0.5
        uc3=zeros(Tsim/Dt,1);    % Control signal array for b=0
        yc1=zeros(Tsim/Dt,1);    % Process output array for b=1
        yc2=zeros(Tsim/Dt,1);    % Process output array for b=0.5
        yc3=zeros(Tsim/Dt,1);    % Process output array for b=0
        rc=zeros(Tsim/Dt,1);     % Reference signal array
        r=0;                     % Initial reference value
        dist=0;                  % Disturbance signal

    case 4 % Example 4 - Feedforward and anti-windup

        %%%%%% Process parameters
        
        % Pu dynamics 
        K=1;        % Static gain
        T=3;        % Time constant
        L=0.5;      % Time delay
        
        % Pd dynamics
        Kd=2;       % Static gain
        Td=1;       % Time constant
        Ld=0.5;     % Time delay
        
        y0=0;       % Initial value of the process ouput
        u0=0;       % Initial value of the controller ouput
        
        umax=7;     % Maximum control limit signal
        umin=0;     % Minimum control limit signal
        dumax=Inf;  % Maximum Increment control limit signal 
        dumin=-Inf; % Minimum Increment control limit signal
                
        Dt=0.01;    % Sampling time        
        
        %%%%%% Simulation paramters
        Tsim=70+L;   % Simulation time  
        
        % Discrete dynamics for simulation of the process model
        
        % Pu 
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Pu
        apd=exp(-Dt/Td); bpd=Kd*(1-apd); Lpd=Ld/Dt;
        
        % Array for simulation
        uff=zeros(Tsim/Dt,1);       % Feedforward control signal array
        uc1=zeros(Tsim/Dt,1);        % Control signal array without saturation - case 1
        uc2=zeros(Tsim/Dt,1);        % Control signal array without anti-windup - case 2
        uc3=zeros(Tsim/Dt,1);        % Control signal array with internal feedforward + clamping - case 3
        uc4=zeros(Tsim/Dt,1);        % Control signal array with internal feedforward + tracking - case 4
        utotal1=zeros(Tsim/Dt,1);   % Total control singal  - case 1
        utotal2=zeros(Tsim/Dt,1);   % Total control singal  - case 2
        utotal3=zeros(Tsim/Dt,1);   % Total control singal  - case 3
        utotal4=zeros(Tsim/Dt,1);   % Total control singal  - case 4
        yc1=zeros(Tsim/Dt,1);        % Process output array - case 1
        yc2=zeros(Tsim/Dt,1);        % Process output array - case 2
        yc3=zeros(Tsim/Dt,1);        % Process output array - case 3
        yc4=zeros(Tsim/Dt,1);        % Process output array - case 4
        y1=zeros(Tsim/Dt,1);        % Total process output Pu+Pd - case 1
        y2=zeros(Tsim/Dt,1);        % Total process output Pu+Pd - case 2
        y3=zeros(Tsim/Dt,1);        % Total process output Pu+Pd - case 3
        y4=zeros(Tsim/Dt,1);        % Total process output Pu+Pd - case 4
        yd=zeros(Tsim/Dt,1);        % Disturbance dynamics array 
        ud=zeros(Tsim/Dt,1);        % Disturbance signal array
        rc=ones(Tsim/Dt,1);         % Reference signal array
        r=0;                  

    case 5 % Example 5 - Filtering measurement noise

        %%%%%% Process parameters
        
        K=1;            % Static gain
        T=1;            % Time constant
        L=0.5;          % Time delay
        
        y0=0;           % Initial value of the process ouput
        u0=0;           % Initial value of the controller ouput
                        
        %%%%%% Simulation paramters
        Tsim=20+L;      % Simulation time  

        Dt=0.01;    % Sampling time        
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        uc1=zeros(Tsim/Dt,1);    % Control signal array for C1
        uc2=zeros(Tsim/Dt,1);    % Control signal array for C2
        uc3=zeros(Tsim/Dt,1);    % Control signal array for C3        
        yc1=zeros(Tsim/Dt,1);    % Process output array for C1
        yc2=zeros(Tsim/Dt,1);    % Process output array for C2
        yc3=zeros(Tsim/Dt,1);    % Process output array for C3        
        rc=zeros(Tsim/Dt,1);     % Reference signal array
        r=0;                     % Initial reference value
        
        % Noise signal
        SNRdb = 5; %Given SNRdb
        variance = 1/(Dt*(10^(SNRdb/10))); 
        load semilla
        rng(s); 
        W = sqrt(variance).*randn(Tsim/Dt); %Gaussian white noise W
        noisesignal=0.01*W;        

    case 6 % Example 6 - Adaptive control of tank level

        %%%% Tank Process parameters
        A=390;a=2.15;g=983;
        
        % Operating point 1
        h10=4;
        T1=(A/a)*sqrt((2*h10/g));   % Time constant
        K1=T1/A;                    % Static gain
        L=0;                        % Time delay
        
        q10=a*sqrt(2*g*h10);        % Starting operating point
        
        % Operating point 2
        h20=12;
        T2=(A/a)*sqrt((2*h20/g));   % Time constant
        K2=T2/A;                    % Static gain
        L=0;                        % Time delay
        
        % Operating point 3
        h30=20;
        T3=(A/a)*sqrt((2*h30/g));   % Time constant
        K3=T3/A;                    % Static gain
        L=0;                        % Time delay       

        %%%%%% Simulation paramters
        Tsim=600;           % Simulation time
        r=h10;              % Initial reference value

        Dt=0.01;    % Sampling time        
        
        % Delay
        Lp=ceil(L/Dt);
        
        % Array for simulation for paralell controllers
        q10=a*sqrt(2*g*h10);
        uc=ones(Tsim/Dt,1)*q10;
        yc=ones(Tsim/Dt,1)*h10;
        rc=ones(Tsim/Dt,1)*h10;
        uc1=ones(Tsim/Dt,1)*q10;
        uc2=ones(Tsim/Dt,1)*q10;
        uc3=ones(Tsim/Dt,1)*q10;
        modes=ones(Tsim/Dt,1);

        % Array for simulation for single controller
        ucs=ones(Tsim/Dt,1)*q10;  
        ycs=ones(Tsim/Dt,1)*h10;
        rcs=ones(Tsim/Dt,1)*h10;       
        modess=ones(Tsim/Dt,1);

        % Global closed-loop time constant
        Tc=(T1+T2+T3)/3;
        
    case 7 % Example 7 - Selectors

        %%%%%% Process parameters
        
        K=1;        % Static gain
        T=1;        % Time constant
        L=0.5;      % Time delay
        
        %%%%%% Simulation paramters
        Tsim=150;   % Simulation time  
        
        Dt=0.01;    % Sampling time
        
        % Discrete dynamics for simulation of the process model
        ap=exp(-Dt/T); bp=K*(1-ap); Lp=L/Dt;
        
        % Array for simulation
        r1=0.3;                  % Initial reference value for r1
        r2=0.5;                  % Initial reference value for r2
        
        uc1=zeros(Tsim/Dt,1);    % Control signal array controller 1
        uc2=zeros(Tsim/Dt,1);    % Control signal array controller 2
        uc=zeros(Tsim/Dt,1);     % Final control signal array 
        yc1=ones(Tsim/Dt,1)*r1;  % Process output array y1
        yc2=ones(Tsim/Dt,1)*r2;  % Process output array y2
        rc1=ones(Tsim/Dt,1)*r1;  % Reference signal for y1
        rc2=ones(Tsim/Dt,1)*r2;  % Reference signal for y2
        
        load v;                  % Disturbance signal profile      

    otherwise % Deafult example 
        error('Wrong Example ID');            
end

File: python/process_and_simulation_data.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import numpy as np
from scipy.io import loadmat


def _n_samples(tsim: float, dt: float) -> int:
    return int(round(tsim / dt))


def _extract_seed_from_mat_s(s_obj: np.ndarray) -> int:
    # semilla.mat guarda el estado de rng de MATLAB como struct.
    try:
        return int(np.asarray(s_obj[0, 0]["Seed"]).squeeze())
    except Exception:
        return int(np.asarray(s_obj).squeeze())


def load_process_and_simulation_data(example_id: int, base_path: str | Path | None = None) -> dict[str, Any]:
    base = Path(base_path) if base_path is not None else Path(__file__).resolve().parent
    data: dict[str, Any] = {"exampleID": example_id}

    if example_id == 0:
        k = 1.0
        t = 1.0
        l = 0.5
        y0 = 0.0
        u0 = 0.0
        tsim = 20.0 + l
        dt = 0.01

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))
        n = _n_samples(tsim, dt)

        uc = np.zeros(n)
        yc = np.zeros(n)
        rc = np.zeros(n)
        r = 0.0

        snr_db = 5.0
        variance = 1.0 / (dt * (10.0 ** (snr_db / 10.0)))
        semilla = loadmat(base / "semilla.mat")
        seed = _extract_seed_from_mat_s(semilla["s"])
        rng = np.random.default_rng(seed)
        w = np.sqrt(variance) * rng.standard_normal(n)
        noise_signal = 0.01 * w

        data.update(locals())

    elif example_id == 1:
        k = 1.0
        t = 1.0
        l = 0.5
        y0 = 0.0
        u0 = 0.0
        tsim = 20.0 + l
        dt = 0.01

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))
        n = _n_samples(tsim, dt)

        uc = np.zeros(n)
        yc = np.zeros(n)
        rc = np.zeros(n)
        r = 3.0

        data.update(locals())

    elif example_id == 2:
        k = 1.0
        t = 1.0
        l = 0.5
        y0 = 0.0
        u0 = 1.0
        tsim = 30.0 + l
        dt = 0.1

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))
        n = _n_samples(tsim, dt)

        uc = np.ones(n)
        yc = np.ones(n)
        rc = np.ones(n)
        r = 1.0

        data.update(locals())

    elif example_id == 3:
        k = 1.0
        t = 1.0
        l = 0.5
        y0 = 0.0
        u0 = 0.0
        tsim = 20.0 + l
        dt = 0.01

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))
        n = _n_samples(tsim, dt)

        uc1 = np.zeros(n)
        uc2 = np.zeros(n)
        uc3 = np.zeros(n)
        yc1 = np.zeros(n)
        yc2 = np.zeros(n)
        yc3 = np.zeros(n)
        rc = np.zeros(n)
        r = 0.0
        dist = 0.0

        data.update(locals())

    elif example_id == 4:
        k = 1.0
        t = 3.0
        l = 0.5

        kd = 2.0
        td = 1.0
        ld = 0.5

        y0 = 0.0
        u0 = 0.0

        umax = 7.0
        umin = 0.0
        dumax = np.inf
        dumin = -np.inf

        dt = 0.01
        tsim = 70.0 + l

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))

        apd = np.exp(-dt / td)
        bpd = kd * (1.0 - apd)
        lpd = int(round(ld / dt))

        n = _n_samples(tsim, dt)

        uff = np.zeros(n)
        uc1 = np.zeros(n)
        uc2 = np.zeros(n)
        uc3 = np.zeros(n)
        uc4 = np.zeros(n)
        utotal1 = np.zeros(n)
        utotal2 = np.zeros(n)
        utotal3 = np.zeros(n)
        utotal4 = np.zeros(n)
        yc1 = np.zeros(n)
        yc2 = np.zeros(n)
        yc3 = np.zeros(n)
        yc4 = np.zeros(n)
        y1 = np.zeros(n)
        y2 = np.zeros(n)
        y3 = np.zeros(n)
        y4 = np.zeros(n)
        yd = np.zeros(n)
        ud = np.zeros(n)
        rc = np.ones(n)
        r = 0.0

        data.update(locals())

    elif example_id == 5:
        k = 1.0
        t = 1.0
        l = 0.5
        y0 = 0.0
        u0 = 0.0
        tsim = 20.0 + l
        dt = 0.01

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))
        n = _n_samples(tsim, dt)

        uc1 = np.zeros(n)
        uc2 = np.zeros(n)
        uc3 = np.zeros(n)
        yc1 = np.zeros(n)
        yc2 = np.zeros(n)
        yc3 = np.zeros(n)
        rc = np.zeros(n)
        r = 0.0

        snr_db = 5.0
        variance = 1.0 / (dt * (10.0 ** (snr_db / 10.0)))
        semilla = loadmat(base / "semilla.mat")
        seed = _extract_seed_from_mat_s(semilla["s"])
        rng = np.random.default_rng(seed)
        w = np.sqrt(variance) * rng.standard_normal(n)
        noise_signal = 0.01 * w

        data.update(locals())

    elif example_id == 6:
        a_tank = 390.0
        a_orifice = 2.15
        g = 983.0

        h10 = 4.0
        t1 = (a_tank / a_orifice) * np.sqrt((2.0 * h10 / g))
        k1 = t1 / a_tank
        l = 0.0
        q10 = a_orifice * np.sqrt(2.0 * g * h10)

        h20 = 12.0
        t2 = (a_tank / a_orifice) * np.sqrt((2.0 * h20 / g))
        k2 = t2 / a_tank

        h30 = 20.0
        t3 = (a_tank / a_orifice) * np.sqrt((2.0 * h30 / g))
        k3 = t3 / a_tank

        tsim = 600.0
        r = h10
        dt = 0.01
        lp = int(np.ceil(l / dt))
        n = _n_samples(tsim, dt)

        uc = np.ones(n) * q10
        yc = np.ones(n) * h10
        rc = np.ones(n) * h10
        uc1 = np.ones(n) * q10
        uc2 = np.ones(n) * q10
        uc3 = np.ones(n) * q10
        modes = np.ones(n)

        ucs = np.ones(n) * q10
        ycs = np.ones(n) * h10
        rcs = np.ones(n) * h10
        modess = np.ones(n)

        tc = (t1 + t2 + t3) / 3.0

        data.update(locals())

    elif example_id == 7:
        k = 1.0
        t = 1.0
        l = 0.5
        tsim = 150.0
        dt = 0.01

        ap = np.exp(-dt / t)
        bp = k * (1.0 - ap)
        lp = int(round(l / dt))

        n = _n_samples(tsim, dt)

        r1 = 0.3
        r2 = 0.5
        uc1 = np.zeros(n)
        uc2 = np.zeros(n)
        uc = np.zeros(n)
        yc1 = np.ones(n) * r1
        yc2 = np.ones(n) * r2
        rc1 = np.ones(n) * r1
        rc2 = np.ones(n) * r2

        v_mat = loadmat(base / "v.mat")
        v = np.asarray(v_mat["v"]).squeeze()
        if v.shape[0] < n:
            pad = np.zeros(n - v.shape[0])
            v = np.concatenate([v, pad])
        else:
            v = v[:n]

        data.update(locals())

    else:
        raise ValueError("Wrong Example ID")

    for key in list(data.keys()):
        if key.startswith("__"):
            data.pop(key, None)
    return data
